### ISSUE TITLE
Single file test targets

### ISSUE THREAD

**Issue by caseyduquettesc** [bazelbuild/bazel-gazelle#1251]
I'm merely posing a question if gazelle has considered a directive to configure single file test targets before? I couldn't find any past issues, but maybe I missed it.

Projects with a mostly flat structure end up with one big test target when using gazelle. When I was maintaining the BUILD.bazel file by hand, I would create separate test targets for each test file so I could take advantage of remote caching and get better flaky test reporting (smaller targets means easier to debug flakes).

I thought it might be nice to have the option of generating a bunch of smaller test targets if the user wanted.

What do you all think? Or please enlighten me if there's a benefit that isn't obvious to me of having a single larger target. Thanks!

### What version of gazelle are you using?

0.25.0

### What version of rules_go are you using?

0.32.0

### What version of Bazel are you using?

5.1.1

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

darwin x86_64

**Comment by achew22** [comment]
Very interesting question. I don't think Gazelle supports this right now, but I am curious about what makes it such a desirable feature. When I'm trying to work on a subset of tests, I usually reach for the [`--test_filter`](https://docs.bazel.build/versions/main/command-line-reference.html#flag--test_filter) option which is implemented in most languages. Doing `bazel test :my_one_file` vs `bazel test --test_filter=one_file :tests` doesn't seem like much of a distinction to me. When it comes to flaky detection, I would reach for something that tracked the output from each individual test case inside of the tests and aggregated that information. I think, maybe, I don't understand your workflow. Could you enlighten me a little bit more?

Thanks!

**Comment by caseyduquettesc** [comment]
I think test filtering does make working with larger targets easier, and that's on me for not using it as much as I could.

I've worked a lot in more modular repositories where I got used to being able to run `test //...` and relying on caching/incremental testing. I'll also have a "run all" configuration in GoLand. Then when I came to a flatter project, if I add a new test file to a package, bazel runs all the tests in that package. It's not the end of the world, I just got spoiled by really fast incremental test times.

You're absolutely right that I could achieve a similar effect with `--test_filter` on the new tests. Ergonomically, I don't always know what test suite I'm working in. We often use suites (from `github.com/stretchr/testify`) to keep the test names from clashing in packages that have a lot of test files. I suppose your argument makes my request much less desirable and is now a quality of life improvement vs a "functionally my incremental tests are slower" :)

Thinking out loud a bit now, I'm wondering what if gazelle could make use of [`test_suite`](https://bazel.build/reference/be/general#test_suite) and broke out the test files into their own targets, but produced another target by the same name it does today (`package_test`), but as a `test_suite`. There would still be a target representing all tests in the package.

**Comment by caseyduquettesc** [comment]
> When it comes to flaky detection, I would reach for something that tracked the output from each individual test case inside of the tests and aggregated that information.

Yeah, I think the flaky test detection shouldn't be at the target level. Let's ignore that one

**Comment by achew22** [comment]
Last time I was using JetBrains there was a very nice button that you could click that was right next to the definition of the test case. You could use that to run just a single test target.

Sorry to push back so hard on this. While it's possible to do this, I think it has a lot of complexity. For example if you have `foo_test.go` and `bar_test.go` and `foo_test.go` defines a `func runFooForTesting(t *testing.T)`, which is a fairly common pattern, we would need to parse out the dependency graph of each testcase inside those `_test.go` files to determine that `bar_test.go`'s target actually depends on `foo_test.go`. That's just for go, which is one of the most regular languages that's out there. I can't even begin to imagine doing this for C++.

I just wanna summarize the problem, as I understand it.

> When using JetBrains GoLand to do development in a Go repo that is built with Bazel, it is too hard to selectively target just the tests that a developer is working on.

Personally, I think the fix for that is to have a conversation with the GoLand team about improving their tooling. I don't think this is something that should be fixed at the build graph level. If you disagree, I'd love to hear some more thoughts on the topic.

**Comment by caseyduquettesc** [comment]
I don't mind the debate. Regarding the problem statement, I don't think this is IDE specific, it was just one use case. When I originally created the issue, I had the command line in mind. I find it easier to type `test //...` all the time and catch all affected tests rather than selecting one. 

Here's a hypothetical I'm hoping exemplifies my point. Let's say I have a shared test fixture library that has this http client constructor. 

```go
package test_http

type RoundTripFunc func(req *http.Request) *http.Response

func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return f(req), nil
}

//NewTestClient returns *http.Client with Transport replaced to avoid making real calls
func NewTestClient(fn RoundTripFunc) *http.Client {
	return &http.Client{
		Transport: RoundTripFunc(fn),
	}
}
```

In a flat-ish project structure, I might have N .go files, but only `serviceA_client.go` and `serviceB_client.go` that both make http requests and need a test client. They would use `NewTestClient` from the project's `test_http` package. 

Let's say this is a project that generates metrics and ships them to two backends. Some other tests might include generating lots of dummy metrics to test the handling of large volume, and a concurrency test to test generating lots of metrics at the same time. These tests in particular are heavy cpu users and maybe I'm on a laptop where they would easily cause the fan to spin up.

If I renamed `NewTestClient` to `CreateTestClient`, I'd want to see if I broke any tests. My current workflow (at least in a larger, modular project) would be to simply run `test //...` and I can trust that any code I changed would get tested. However, if all the tests were in the same package, I'd have to run every test in the package, including those cpu heavy tests even though they were not impacted by the change to `NewTestClient`. I'd expect to be able to run tests for `serviceA_client_test.go` and `serviceB_client_test.go` without knowing those are the two test files that depended on `NewTestClient`.

To your point though, I do see how even if we wanted to solve this, it's more difficult than it initially seemed to me because gazelle would need to understand the dependencies between files within packages, and not just the packages. Thanks for explaining that.

Given the high effort required and low impact this would have, I'd be okay closing this. At least we have documentation now the idea was visited :)
